### PR TITLE
allow unrestricted reads from zookeeper

### PR DESCRIPTION
**Pull request by serxa**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Now you can read `system.zookeeper` table without restrictions on path or using `like` expression. This reads can generate quite heavy load for zookeeper so to enable this ability you have to enable setting `allow_unrestricted_reads_from_keeper`.

### CODE REVIEW

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 1]
```diff
@@ -176,18 +176,56 @@ static Paths extractPath(const ASTPtr & query, ContextPtr context)
     return extractPathImpl(*select.where(), res, context) ? res : Paths();
 }
 
+static void addRow(MutableColumns & res_columns, const String& node, const String& path, const Coordination::GetResponse& res)
+{
+    if (res.error == Coordination::Error::ZNONODE)
```
Comment (kitaisreal) [inline src/Storages/System/StorageSystemZooKeeper.cpp:181]: From my perspective it is better to move this check into method caller, that way method `addRow` and change argument from `Coordination::GetResponse` to `Coordination::Stat`, that way method will be responsible only for inserting statistics into result columns.

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 2]
```diff
@@ -176,18 +176,56 @@ static Paths extractPath(const ASTPtr & query, ContextPtr context)
...
+        {
+            queue.push_back(path);
+        }
+    }
+
     std::unordered_set<String> paths_corrected;
-    for (const auto & path : paths)
+    while (!queue.empty())
     {
+        const String path = queue.front();
```
Comment (kitaisreal) [inline src/Storages/System/StorageSystemZooKeeper.cpp:226]: We can move path `std::move(queue.front())`.

**Review by kitaisreal — APPROVED** [review]
Pointed to some minor style changes

**Comment by serxa** [comment]
fixes #25645

**Review by kitaisreal — APPROVED** [review]

**Review by kitaisreal — APPROVED** [review]

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 3]
```diff
@@ -180,14 +181,34 @@ static Paths extractPath(const ASTPtr & query, ContextPtr context)
...
+            queue.push_back("/");
+        else
+            throw Exception("SELECT from system.zookeeper table must contain condition like path = 'path' or path IN ('path1','path2'...) or path IN (subquery) in WHERE clause.", ErrorCodes::BAD_ARGUMENTS);
+    }
+    else
+    {
+        recursive = false; // do not recurse if path was specified
+        for (const auto & path : paths)
+        {
+            queue.push_back(path);
```
Comment (alesapin) [inline src/Storages/System/StorageSystemZooKeeper.cpp:202]: reserve before `push_back`?

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 4]
```diff
@@ -228,6 +249,11 @@ void StorageSystemZooKeeper::fillData(MutableColumns & res_columns, ContextPtr c
             res_columns[col_num++]->insert(stat.pzxid);
             res_columns[col_num++]->insert(
                 path); /// This is the original path. In order to process the request, condition in WHERE should be triggered.
+
+            if (recursive && res.stat.numChildren > 0)
+            {
+                queue.push_back(path_part + '/' + nodes[i]);
```
Comment (alesapin) [inline src/Storages/System/StorageSystemZooKeeper.cpp:255]: `fs::path(path_part) / nodes[i]`

**Review by alesapin — COMMENTED** [review]
Need a test.

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 5]
```diff
@@ -180,14 +181,34 @@ static Paths extractPath(const ASTPtr & query, ContextPtr context)
...
+    if (paths.empty())
+    {
+        if (context->getSettingsRef().allow_unrestricted_reads_from_keeper)
+            queue.push_back("/");
+        else
+            throw Exception("SELECT from system.zookeeper table must contain condition like path = 'path' or path IN ('path1','path2'...) or path IN (subquery) in WHERE clause.", ErrorCodes::BAD_ARGUMENTS);
+    }
+    else
+    {
+        recursive = false; // do not recurse if path was specified
```
[the remark this answers is not in the document]
Comment (serxa) [reply src/Storages/System/StorageSystemZooKeeper.cpp:199]: I was thinking about it, but I don't know good way to do it. If query with WHERE path = '/some/path' we will recursively return rows with different value of path... This looks contradictory to me.

[the remark this answers is not in the document]
Comment (serxa) [reply src/Storages/System/StorageSystemZooKeeper.cpp:199]: Probably we can add another code route for `WHERE path LIKE '/some/path%'` What do you think about it?

[the remark this answers is not in the document]
Comment (alesapin) [reply src/Storages/System/StorageSystemZooKeeper.cpp:199]: Really good idea.

**Review by alesapin — APPROVED** [review]

**Comment by alesapin** [comment]
Only flaky check important

**Comment by alesapin** [comment]
Stateless timeout failures Ok.

**Pull request merged**